### DISCUSSION TITLE
USB issues

### DISCUSSION THREAD

**Discussion by brentward** [picocomputer/community#26] [Q&A] [answered]
I have a newly assembled RP6502 which is mostly working as expected but I've had a few issues with USB.

First, depending on which ports on my USB hub I use some devices will not be recognized. For example, I will have 1 USB drive and 1 keyboard on the hub. When I run the status command I will get: `USB : 2 devices` but under that it will only list the keyboard (sometimes it is the keyboard that isn't recognized but the drive is). I have tried a few different drives, keyboards, USB hubs, and USB OTG devices and all of them exhibit this behavior.  Has anyone else had this issue? Sometimes the devices will be recognized after running the `reboot` command but not always.
Are the USB devices supposed to be hot swappable?
Also some of my keyboards will display an error after every key press:
 `Couldn't find the report info for this report !`
They will still work but they are not usable because of the excess messages. If I run status with these keyboards attached they are listed as: 
`HID keyboard and 2 other reports` Could this be an error level setting for the RIA?

**Comment by rumbledethumps** [comment, marked as answer]
I removed the "Couldn't find..." message; it was a placeholder for figuring out joysticks which none of my keyboards triggered. I regularly swap USB drives but you can't swap SD cards in a reader (yet). The TinyUSB and RP6502 projects need contributors who know USB at a very low level.

**Comment by brentward** [comment]
I built the RIA firmware using the latest in the vga branch and that did indeed make the message go away. 
Also, on this build so far the USB issues seemed to have gone away. I can now hot swap the drives and all the ports on my hub seem to be consistently working.

BTW, I also replaced the vga firmware with the build from the latest vga branch but that doesn't seem likely to have affected the USB issues.

Side question: I built this as a debug build, a release build and a release build with debug symbols but for all three the ria uf2 file was significantly smaller than the one included in the zip file. The zip file version is 1255 KB and all of mine have been around 210 -250 KB (even the main branch). The vga uf2 is about the same size. Is my firmware being built incorrectly or has the size significantly changed or is something else going on?

**Reply by rumbledethumps** [reply]
Full builds include all the code pages.
https://github.com/picocomputer/rp6502/blob/f576d72890658408bfeb865f67175c6cc222d663/src/CMakeLists.txt#L155

**Comment by midicdj1000** [comment]
What hub do you recommend ?

**Comment by midicdj1000** [comment]
So far I have tried the Pi keyboard with hub and that works ok including a basic usb storage device, but no controllers worked. Also the caps lock  is not working at all. Next I tried an old mister otg hub lead  that was very cheap with a usb keyboard and this did not work at all. The keyboard did work with a otg - usb converter, but again no caps lock function. I am going to dig out my old Mister powered usb hub and give that a try. So the caps lock is a firmware issue ?

**Reply by rumbledethumps** [reply]
Many things don't work yet. My focus so far has been to validate the hardware and get it in the hands of people who can help with software or spread the word.

**Reply by midicdj1000** [reply]
I have lots of 6502 coder friends and I am trying to make a video to show the kit.

**Comment by Nordlaw-tech** [comment]
Does anyone know which USB hubs do work? So far I have failed to find one that does, but several that don't.

**Reply by midicdj1000** [reply]
this is working for me
https://amzn.to/3YMlLpT
also the pi keyboard

**Reply by brentward** [reply]
I've tried two and had issues with both where depending on which port I use the USB won't mount the file system.
This one works more reliably: https://www.amazon.com/gp/product/B01HYJLZH6/ref=ppx_yo_dt_b_asin_title_o02_s01?ie=UTF8&psc=1
I can also use this: https://www.amazon.com/gp/product/B00L2442H0/ref=ppx_yo_dt_b_asin_title_o08_s00?ie=UTF8&th=1 with this OTG adapter: https://www.amazon.com/gp/product/B083KJ1B5H/ref=ppx_yo_dt_b_asin_title_o07_s01?ie=UTF8&psc=1

Perhaps try switching around which ports you are plugging into with the hubs you do have.

**Comment by Nordlaw-tech** [comment]
Thanks Lee - I've ordered one.

Regards

Geoff

----------------------------------------

**Comment by Nordlaw-tech** [comment]
Thanks for your help.

regards

Geoff

----------------------------------------

**Comment by vruumllc** [comment]
My favorite ANKER wireless keyboard mostly works, but generates "(0 0 0) ---" messages quite often on the screen.
I don't see this using my other wireless keyboard -- a Logitech K470 --  but numlock and capslock don't work.

**Reply by vruumllc** [reply]
Anker (0,0,0) --- message is no longer an issue with the code pushed on 10/14. Yay!!
CapLock also works on both my Anker and Logitech keyboards. 
The NumLock key on Logitech keyboard still doesn't work, but no big deal (Anker has no numeric keypad).

**Comment by RussellSenior** [comment]
I have had major slowness (~200 bytes per second loads) issues with a couple HP branded USB2  thumb drives (v115w, 4GB and v165w 8GB) and a generic I got as swag at a conference. I eventually found a USB-SD card adapter that worked "normally". Also a SanDisk Cruzer Glide 16GB works. I am wondering if people can report which USB drives worked for them. 

doesn't work:
```
03f0:5307 HP, Inc v165w Stick
03f0:3d07 HP, Inc v115w
27c6:609c Shenzhen Goodix Technology Co.,Ltd. Goodix USB2.0 MISC
```
works:
```
05e3:0749 Genesys Logic, Inc. SD Card Reader and Writer
0781:5575 SanDisk Corp. Cruzer Glide
0781:5597 SanDisk Corp.  SanDisk 3.2Gen1
abcd:1234 LogiLink UDisk flash drive (from Multnomah County Library, obviously bogus IDs)
```
